dockstore-webservice/src/main/resources/openapi3/openapi.yaml

```diff
@@ -839,6 +841,43 @@ paths:
       - bearer: []
       tags:
       - tokens
+  /categories:
+    get:
+      description: Retrieve all categories.
+      operationId: getCategories
+      responses:
+        default:
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Category'
+          description: default response
+      summary: Retrieve all categories.
+      tags:
+      - categories
+  /categories/name/{name}:
```

Comment:
these two category endpoints can probably be collapsed. they roughly return they same response. should use a query parameter for filtering. https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/#filtering

Reply:
The two endpoints currently return different types though. How do you propose to resolve that?

1. Have the Java method return a Response? Then the Swagger doesn't the type
2. Have this method return a Category, but with only name field populated?

I'm OK with 2, don't really like 1... Or is there some other idea you have?

Reply:
Ignore my comment, didn't look closely and I thought this was in reference to a `category/names` endpoint.

Reply:
These are the two /categories endpoints:

/categories
/categories/name/{name}

The first returns an array of all Categories, the other returns a single Category with the given name.  Since the return types are different, does it make sense to unite them?

Reply:
yes, the return type isn't different enough.  vast majority of restful apis query faram filter and can return one item.

Reply:
If the idea is to get single category, I think the correct RESTful pattern is `/categories/<identifier>`. The fact that there are two unique identifiers for the category, name and id, complicates things.

I would tend to go with `/categories/<name>`, but I have seen your current proposal when there is more than one unique id. But since we're not surfacing `/categories/<id>`, I would lean towards `categories/<name>`.

Reply:
I modeled the naming conventions of the /categories endpoints after some of the other existing endpoints.  It'd probably be good to stick with a common scheme.  For example, here's some existing organizations endpoints:

/organizations
/organizations/{organizationId}
/organizations/name/{name}

We can add an endpoint to retrieve a category by id, it's trivial.

Reply:
some of the existing endpoints aren't restful which is something we've been trying to move towards. i would not take inspiration from existing endpoints.

the `categories/<name>` is also something i've considered but i find it unusual for it to not be `id`

common scheme should be REST

My first comment already linked a source. He's a 2nd one https://restfulapi.net/resource-naming/ (see bottom of it)  . feel free to link your own sources too

Reply:
Since, in the case of Categories, the 'name' is essentially a unique id space, and because names start with a letter, disjoint from the numeric category id space, perhaps the most RESTful set of endpoints, from a purely philosophical view, is:

/categories
/categories/{id_or_name}

The latter endpoint retrieves by numeric id if a number is specified, and by name otherwise.
(this is not a recommendation)

Reply:
I don't think I've seen something like `/categories/{id_or_name}` as a best practice before. as mentioned, if you can find some sources for this for this or some other alternative, that'd be great.

Reply:
This stackoverflow post relates to our question, enumerates some naming approaches, and raises some objections to using a query string on a "list of" endpoint to retrieve a single resource by a unique name:
https://stackoverflow.com/questions/41103091/rest-api-design-with-multiple-unique-ids
We're using approach #4 in the PR to retrieve a category by name.
Here's another, the second answer is the current scheme:
https://stackoverflow.com/questions/842261/rest-supporting-multiple-possible-identifiers